build/gen_stub.php

```diff
@@ -177,88 +230,164 @@ public static function void(): SimpleType
-        case "object":
+            case "null":
+                return "MAY_BE_NULL";
+            case "false":
+                return "MAY_BE_FALSE";
+            case "bool":
+                return "MAY_BE_BOOL";
+            case "int":
+                return "MAY_BE_LONG";
+            case "float":
+                return "MAY_BE_DOUBLE";
+            case "string":
+                return "MAY_BE_STRING";
+            case "array":
+                return "MAY_BE_ARRAY";
+            case "object":
+                return "MAY_BE_OBJECT";
+            case "callable":
+                return "MAY_BE_CALLABLE";
+            case "mixed":
+                return "MAY_BE_ANY";
+            case "static":
+                return "MAY_BE_STATIC";
+            case "never":
+                return "MAY_BE_NEVER";
+            default:
+                throw new Exception("Not implemented: $this->name");
+        }
+    }
+
+    public function toOptimizerTypeMaskForArrayKey(): string {
+        assert($this->isBuiltin);
+
+        switch (strtolower($this->name)) {
+            case "int":
+                return "MAY_BE_ARRAY_KEY_LONG";
+            case "string":
+                return "MAY_BE_ARRAY_KEY_STRING";
+            default:
+                throw new Exception("Type $this->name cannot be an array key");
+        }
+    }
+
+    public function toOptimizerTypeMaskForArrayValue(): string {
+        if (!$this->isBuiltin) {
+            return "MAY_BE_ARRAY_OF_OBJECT";
+        }
+
+        switch (strtolower($this->name)) {
+            case "null":
+                return "MAY_BE_ARRAY_OF_NULL";
+            case "false":
+                return "MAY_BE_ARRAY_OF_FALSE";
+            case "bool":
+                return "MAY_BE_ARRAY_OF_FALSE|MAY_BE_ARRAY_OF_TRUE";
+            case "int":
+                return "MAY_BE_ARRAY_OF_LONG";
+            case "float":
+                return "MAY_BE_ARRAY_OF_DOUBLE";
+            case "string":
+                return "MAY_BE_ARRAY_OF_STRING";
+            case "array":
+                return "MAY_BE_ARRAY_OF_ARRAY";
+            case "object":
+                return "MAY_BE_ARRAY_OF_OBJECT";
+            case "resource":
+                return "MAY_BE_ARRAY_OF_RESOURCE";
+            case "mixed":
+                return "MAY_BE_ARRAY_OF_ANY";
+            default:
+                throw new Exception("Type $this->name cannot be an array value");
+        }
+    }
+
+    public function toOptimizerTypeMask(): string {
+        if ($this->isBuiltin && strtolower($this->name) === "resource") {
+            return "MAY_BE_RESOURCE";
+        }
+
+        if ($this->isBuiltin && strtolower($this->name) === "true") {
+            return "MAY_BE_TRUE";
+        }
+
+        if ($this->isBuiltin && strtolower($this->name) === "mixed") {
+            return "MAY_BE_ANY|MAY_BE_ARRAY_KEY_ANY|MAY_BE_ARRAY_OF_ANY";
+        }
+
+        if (!$this->isBuiltin) {
```

Comment:
Move this to the start so you don't have to check isBuiltin for the other cases?